### ISSUE TITLE
Multiple extrusions/sweep question for gear generation

### ISSUE THREAD

**Issue by wentam** [CadQuery/cadquery#676]
I'm attempting to port my gear generator from openscad that can produce arbitrary helix patterns, like this:

![image](https://user-images.githubusercontent.com/901583/110029325-55ee0180-7d02-11eb-928a-8d8d3081a782.png)
![image](https://user-images.githubusercontent.com/901583/110029465-80d85580-7d02-11eb-88db-dab0eebc0941.png)

Getting the general gear profile generation working was simple, and splines are awesome:
![image](https://user-images.githubusercontent.com/901583/110029741-d6acfd80-7d02-11eb-8b8b-aa851ef71882.png)

But I'm having a hard time finding a good solution to set up the helix generation.

I would expect to be able twistExtrude multiple times (`p=p.wire().twistExtrude(5,20).wires(">Z").twistExtrude(5,-20)`), but there's something I don't understand because this just errors out. How would I extrude the top face of an object?

I had a little bit of success with sweep and an auxSpine
![image](https://user-images.githubusercontent.com/901583/110030358-9dc15880-7d03-11eb-96ae-fa63a517c691.png)

Thats very satisfying, but trying to define a herringbone profile with a polyline as the auxSpine just does strange things
![image](https://user-images.githubusercontent.com/901583/110030400-aade4780-7d03-11eb-9f07-c89a915a9c91.png)

Source is here: https://github.com/wentam/Compound-planetary-gearbox-generator/blob/db55c7ff6a5a3625f614d7179e7d08a9f393230f/test.py

I'd any appreciate ideas on how to best solve this!

**Comment by fedorkotov** [comment]
The method you are missing is [toPending(..)](https://cadquery.readthedocs.io/en/latest/classreference.html?highlight=toPending#cadquery.Workplane.toPending). But it seems that there is a bug in `twistExtrude(..)`.

This works
```python
result = \
    cq.Workplane('XY')\
      .rect(5,5)\
      .twistExtrude(5,20)\
      .wires(">Z")\
      .toPending()\
      .extrude(5)
```
![image](https://user-images.githubusercontent.com/1106650/110076379-731fe000-7d95-11eb-9600-ac4f307c1ccd.png)


But this does not
```python
result = \
    cq.Workplane('XY')\
      .rect(5,5)\
      .twistExtrude(5,20)\
      .wires(">Z")\
      .toPending()\
      .twistExtrude(5,-20)      
      # same with .twistExtrude(-5,-20)
```
Both variants (with distance 5 and -5) extrude backwards and give this shape
![image](https://user-images.githubusercontent.com/1106650/110076318-54b9e480-7d95-11eb-9213-d46f3053a616.png)

**Comment by fedorkotov** [comment]
As a workaround you can probably use [mirror(..)](https://cadquery.readthedocs.io/en/latest/classreference.html?highlight=mirror#cadquery.Workplane.mirror)
```python
result = \
    cq.Workplane('XY')\
      .polygon(6, 5.0)\
      .twistExtrude(5,60)\
      .mirror(union=True)
```
![image](https://user-images.githubusercontent.com/1106650/110077028-87181180-7d96-11eb-9bf6-9ed2ffc7d6cd.png)

**Comment by wentam** [comment]
Thanks! toPending() indeed allows for non-twisted extrusion:

![image](https://user-images.githubusercontent.com/901583/110077950-071d8780-7d55-11eb-9590-a7d0bb0ecd2f.png)

Mirror doesn't fully achieve what is needed here, as asymmetric profiles need to be possible. It is a valid workaround for symmetric cases though.

I don't fully understand what pending vs non-pending wires are, and I can't find much information about it in the docs. Perhaps a candidate for the 'CadQuery Concepts' section?

**Comment by fedorkotov** [comment]
Asymmetric ones you can just glue together from two parts with [union](https://cadquery.readthedocs.io/en/latest/classreference.html?highlight=union#cadquery.Workplane.union) like this

```python
result = \
    cq.Workplane('XY')\
      .polygon(6, 5.0)\
      .twistExtrude(-5,60)\
      .union(
          cq.Workplane('XY')\
            .polygon(6, 5.0)\
            .twistExtrude(5,-120))
```
![image](https://user-images.githubusercontent.com/1106650/110078804-54bbe380-7d99-11eb-8f19-5cd4c57ddab4.png)

**Comment by wentam** [comment]
Yup! - was just playing with union a bit as a workaround for this.

Thanks for your time! That'll be a good solution for now.

**Comment by adam-urbanczyk** [comment]
@wentam great project! As already discussed `toPending` is the way. Don't forget to use `workplane` too:
```python
result = (
    cq.Workplane()
      .rect(5,5)
      .twistExtrude(5,20)
      .wires(">Z").toPending()
      .workplane()
      .twistExtrude(5,-20)
)
```
![image](https://user-images.githubusercontent.com/13981538/110079936-0f43ea00-7d8a-11eb-9d87-84597496f7d0.png)

@fedorkotov see above, you need to use `workplane` to make `twistExtrude` wokr.

**Comment by wentam** [comment]
Aha: workplane does make it behave as expected:
![image](https://user-images.githubusercontent.com/901583/110080208-2833a780-7d58-11eb-939e-a18a8ce2be5e.png)

Makes sense I supposed if extrusion is intended to occur relative to the workplane.

**Comment by adam-urbanczyk** [comment]
Regarding `auxSpine`: probably only C1 curves are supported. I did not check it though

**Comment by wentam** [comment]
@adam-urbanczyk the auxSpine solution doesn't seem to result in usable gears anyway, as the teeth on the opposite side are of a different shape than the front:

![image](https://user-images.githubusercontent.com/901583/110080785-f96a0100-7d58-11eb-8a98-2268f198381f.png)
vs
![image](https://user-images.githubusercontent.com/901583/110080797-fe2eb500-7d58-11eb-9be9-9f4bf0517142.png)

**Comment by adam-urbanczyk** [comment]
AFAIK: your spine needs to be placed on a cylinder for this to work properly. You can take a look at how `twistExtrude` is actually implemented.

**Comment by fedorkotov** [comment]
Thanks a lot @adam-urbanczyk . Issues section of this project is a great supplement for documentation.

**Comment by adam-urbanczyk** [comment]
Yes, I try to add the `question` label for this kind of issues. One day we might have better docs and maybe even a FAQ.

**Comment by Jojain** [comment]
I am also working on gear and I made one here where you can specify the helix angle you want, (even though helix angle >3° make the filleting not working) It's still kindof WIP but maybe can give you some inspiration.
[https://github.com/CadQuery/cadquery-contrib/blob/master/examples/cylindrical_gear.py]
